### PR TITLE
Use notifications to allow/refuse location

### PR DESCRIPTION
**Pull request by ifaresse**
## 🔄 Changes Overview

This PR introduces notifications for location sharing requests, local storage of friends triggering notifications, and tests for related functionality.

---

### Notifications
- **Location Sharing Requests**: Launched a notification for each user, allowing them to accept or refuse location sharing.

### Local Storage
- **Friend Notifications**: Stored the list of friends that triggered notifications locally for better tracking and user experience.

---

## 🧪 Testing

- **`saveNotifiedFriends` and `getSavedAlreadyNotifiedFriends`**: Added tests to ensure the proper saving and retrieval of friends that have already triggered notifications.
- **`ActionReceiver`**: Tested the receiver logic to validate actions triggered by notifications.

---

## 🛠️ Technical Implementation

- Integrated notification handling for location sharing.
- Implemented local storage with json for friends who already notified .

### CODE REVIEW

File: app/src/main/java/com/swent/suddenbump/model/user/UserRepositoryFirestore.kt [hunk 1]
```diff
@@ -1071,6 +1073,27 @@ class UserRepositoryFirestore(private val db: FirebaseFirestore, private val con
     return sharedPreferences.getString("userId", null) ?: ""
   }
 
+  override fun saveNotifiedFriends(friendsUID: List<String>) {
+    val gson = Gson()
+    val jsonString = gson.toJson(friendsUID) // Convert list to JSON string
+    sharedPreferences.edit().putString("notified_friends", jsonString).apply()
```
Comment (marinovdeyan) [inline app/src/main/java/com/swent/suddenbump/model/user/UserRepositoryFirestore.kt:1079]: very good use of this feature ;)

File: app/src/main/java/com/swent/suddenbump/ui/map/Map.kt [hunk 2]
```diff
@@ -151,35 +153,71 @@ fun fetchLocationToServer(location: Location, userViewModel: UserViewModel) {
...
+          PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
+
+  // Build the notification with actions
   val notificationBuilder =
       NotificationCompat.Builder(context, channelId)
           .setSmallIcon(android.R.drawable.ic_dialog_info)
           .setContentTitle("Friend Nearby")
-          .setContentText("A friend is within your radius!")
+          .setContentText(
+              "${friend.firstName} ${friend.lastName} is within your radius! \n Would you like to share your location with them?")
```
Comment (marinovdeyan) [inline app/src/main/java/com/swent/suddenbump/ui/map/Map.kt:215]: This is much better than how it was before. We can now finally know who the notification is about.

**Review by marinovdeyan — APPROVED** [review]
Overall a very good PR with high quality code and extensive usage of multiple Android libraries. LGTM, OK to merge from me.

To be fixed for future PR: when receiving multiple notifications, it sometimes happens that the the Accept/Reject buttons for one notification trigger another notification instead. The bug is minor, but should be fixed in future sprints to ensure full functionality of this feature.

**Comment by ifaresse** [comment]
> Overall a very good PR with high quality code and extensive usage of multiple Android libraries. LGTM, OK to merge from me.
> 
> To be fixed for future PR: when receiving multiple notifications, it sometimes happens that the the Accept/Reject buttons for one notification trigger another notification instead. The bug is minor, but should be fixed in future sprints to ensure full functionality of this feature.

Yeah thanks for your review I'll fix that as soon as possible.

**Pull request merged**